### PR TITLE
🐛   Source Shopify: Improve SubStream state handling

### LINKED ISSUE #10595: source shopify: v0.1.26 stream rename causing child stream to run before parent
## Environment
- **Airbyte version**:  .35.11 & .35.12
- **OS Version / Instance**: ubuntu on ec2
- **Deployment**: docker
- **Source Connector and version**: shopify 0.1.26
- **Severity**: High
- **Step where error happened**: setup new connection through UI, edit connection through UI. create/update over API not affected

## Current Behavior
It seems like streams are read in the order they appear in the sync catalog. The web ui sorts them and `order_refunds` now appears before `orders` after the stream rename, so if you create a new connection or save an edit to one through the web UI it causes that child stream to run before its parent.

Here is an example of the performance change we saw: the sync at the bottom was running refunds before orders, and the sync at the top is running orders before refunds after I fixed the order with a `connections/update` call.
image (9)

Slack thread:
https://airbytehq.slack.com/archives/C01MFR03D5W/p1645646538929519

## Expected Behavior
Connector should probably have some way to specify stream order so that the catalog order doesn't matter. 

## Steps to Reproduce
1. create a new shopify connection, or edit an existing one that was created with orders ahead of refunds in the catalog using the web ui
2. run a sync, you should see the refunds stream start before its parent

### PR DESCRIPTION
**Pull request by bazarnov**
## What
Resolving: https://github.com/airbytehq/airbyte/issues/10595

**Before this update:**
Now each stream could be synced independently from parent stream, for instance:
`Order_refunds` is the sub-stream of the  `Orders` stream.
To have it properly synced with the latest state of the parent, previously you have to choose the 2 streams to sync `Order_refunds` and `Orders` to have the `Orders` stream state, and based on that sync the `Order_refunds` incrementally. Sometimes this situation causes the sub-stream to sync before the parent stream is synced first, which causes to full-refresh instead of incremental one.

**WIth this update:**
You just now choose the `Order_refunds` stream alone and it would sync independently from `Orders` stream state.
Conclusion: each stream now has it's own state, and for parent-dependent streams the state object looks like:
```
{
      "order_refunds": {
          "created_at": "2022-03-03T03:47:45-08:00",
          "orders": {
              "updated_at": "2022-03-03T03:47:46-08:00"
          }
      }
}
```

## How
- re-factored the `ShopifySubStream` class to handle the parent state independently for each SubStream.
- added `unit_tests`

## 🚨 User Impact 🚨
The usual update doesn't require user actions with connector, except of the update itself. 
In case of the issues - downgrade to the previous version is possible as well.

**IMPORTANT NOTE:** 
- The very first sync after the update would be like the Initial Sync as the connector needs to update it's state up to the new state structure. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 1]
```diff
@@ -91,8 +91,19 @@ def state_checkpoint_interval(self) -> int:
     # Setting the default cursor field for all streams
     cursor_field = "updated_at"
 
+    @property
+    def comparison_value(self) -> Union[int, str]:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:95]:
```suggestion
    def default_comparison_value(self) -> Union[int, str]:
```

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py [hunk 2]
```diff
@@ -117,50 +128,56 @@ def filter_records_newer_than_state(self, stream_state: Mapping[str, Any] = None
...
 
-    parent_stream_class: object = None
+    shopify_stream_class: object = None
     slice_key: str = None
     nested_record: str = "id"
     nested_record_field_name: str = None
     nested_substream = None
 
+    @property
+    def shopify_stream(self) -> object:
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-shopify/source_shopify/source.py:154]: woudl this be a better name? 
```suggestion
    def parent_stream(self) -> object:
```

**Review by sherifnada — APPROVED** [review]

File: docs/integrations/sources/shopify.md [hunk 3]
```diff
@@ -100,6 +100,7 @@ This is expected when the connector hits the 429 - Rate Limit Exceeded HTTP Erro
 
 | Version | Date | Pull Request | Subject |
 | :--- | :--- | :--- | :--- |
+| 0.1.35 | 2022-03-07 | [10915](https://github.com/airbytehq/airbyte/pull/10915) | Implemented independent state for each ShopifySubStream |
```
Comment (sherifnada) [inline docs/integrations/sources/shopify.md:103]: framing it in a way a non-engineer can understand ;) 

```suggestion
| 0.1.35 | 2022-03-07 | [10915](https://github.com/airbytehq/airbyte/pull/10915) | Fix a bug which caused full refresh syncs of child REST entities configured for incremental |
```

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 1 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py:95]: Done.

File: airbyte-integrations/connectors/source-shopify/source_shopify/source.py (hunk 2 above)
Comment (bazarnov) [reply airbyte-integrations/connectors/source-shopify/source_shopify/source.py]: Changed. Done.

File: docs/integrations/sources/shopify.md (hunk 3 above)
Comment (bazarnov) [reply docs/integrations/sources/shopify.md:103]: Acceptable. Thank you

**Pull request merged**